src/bot/scripts/card.js

```diff
@@ -728,946 +66,1233 @@ const card = new Command({
+                  .reduce((a, c) => a + c, 0);
+              }
+
+              const addReduceSlot = (a, c) =>
+                a[c.property] == null ? {
+                  ...a,
+                  [c.property]: c.value
+                } : {
+                  ...a,
+                  [c.property]: a[c.property] + c.value
+                };
+
+              const alteredStats = card.tech_tree.slots
+
+                .slice(0, requiredUpgrades)
+                .reduce(
+                  addReduceSlot,
+                  card.tech_tree.levels
+                  .slice(0, requiredLevels)
+                  .reduce((a, c) => [...a, ...c.slots], [])
+                  .reduce(addReduceSlot, {})
+                );
+
+              // stat merge
+              // deep copy cos gay things
+              let alteredCard = JSON.parse(JSON.stringify(card));
+
+              if (alteredStats.power_unlock != null) {
+                alteredCard.is_power_locked = false;
+                delete alteredStats.power_unlock;
+              }
+
+              alteredCard = [alteredStats].reduce((a, c) => {
+                for (let [k, v] of Object.entries(c)) {
+                  if (k.startsWith("stat_")) {
+                    if (a[k.slice(5)] != null) {
+                      a[k.slice(5)] += v;
+                    } else if (a[k.slice(9)] != null) {
+                      a[k.slice(9)] += v;
+                      9;
+                    } else {
+                      return console.error(
+                        "error applying upgrade stats 1: " + k
+                      );
+                    }
+                  } else if (k.startsWith("power_")) {
+                    const powerIndex = a.powers.findIndex(e => e.type === k);
+
+                    if (powerIndex !== -1) {
+                      a.powers[powerIndex].amount += v;
+                    } else {
+                      // if power duration is specified, assume
+                      //  that there is only one power
+                      if (k === "power_duration") {
+                        a.powers[0].duration += v;
+                      } else if (
+                        k === "power_range"
+                      ) {
+                        a.powers[0].radius += v;
+                      } else {
+                        return console.error(
+                          "error applying upgrade stats 2: " + k
+                        );
+                      }
+                    }
+                  } else {
+                    return console.error("error applying upgrade stats 3");
+                  }
+                }
+
+                return a;
+              }, alteredCard);
+
+              alteredCard.description = alteredCard.description.replace(
+                /\{(.*?)\}/g,
+                match => {
+                  const bracketless = match.slice(1, match.length - 1);
+
+                  function getPowerAmount(powerType) {
```

Comment:
global scope, arrow fn

Reply:
this one needs to be inside the _calculateCardAugmentData to acess alteredCard

Reply:
make alteredCard an extra param on the function then, if youre editing the alteredCard in the fn, make sure it either is not kept between script runs or deep copy it

Reply:
ok